x/sync/manager.go

```diff
@@ -584,11 +590,11 @@ func (m *Manager) setError(err error) {
 
 // Mark the range [start, end] as synced up to [rootID].
 // Assumes [m.workLock] is not held.
-func (m *Manager) completeWorkItem(ctx context.Context, work *workItem, largestHandledKey []byte, rootID ids.ID, proofOfLargestKey []merkledb.ProofNode) {
+func (m *Manager) completeWorkItem(ctx context.Context, work *workItem, largestHandledKey merkledb.Maybe[[]byte], rootID ids.ID, proofOfLargestKey []merkledb.ProofNode) {
 	// if the last key is equal to the end, then the full range is completed
-	if !bytes.Equal(largestHandledKey, work.end) {
+	if work.end.IsNothing() || !bytes.Equal(largestHandledKey.Value(), work.end.Value()) {
```

Comment:
Is this correct? We're checking here whether we're done fetching the range in `work`, and if there's no upper bound on `work` (i.e. `work.end.IsNothing()`) then we always treat it as though we're not done fetching the range. Don't we also need to look at `largestHandledKey`? See [here](https://github.com/ava-labs/avalanchego/pull/1657/files#diff-69e61c5702db139bb75c3063cfee14dd84f123d2febf213b2a8767f368ff1664R617)

Reply:
I'm not sure I understand.  Since end is unbound, largestHandledKey will always be unequal which means we won't know if the range is complete just via comparison.  We definitely need to use the findnextkey logic to determine range completion.

Reply:
> Since end is unbound, largestHandledKey will always be unequal

I don't think this is true, actually. It's possible that `work.end.IsNothing() && largestHandledKey.IsNothing()`. Namely, if we request a range/change proof with no upper bound and receuve a range/change proof with no key-values.

That said, I think it might be a bit more readable to change this to:

```go
if !merkledb.MaybeBytesEquals(largestHandledKey, work.end) {
```

as in the code linked above. I just think it's easier to reason about this with "the largest key we handled wasn't equal to the one we wanted", which is what the code above basically says, rather than:

```
if work.end.IsNothing() || !bytes.Equal(largestHandledKey.Value(), work.end.Value()) {
```

which I think is somewhat less clear.